drud/ddev
# Validate config before writing .yaml or starting project, resolves #1095

## The Problem/Issue/Bug:
Only certain config values were being validated before attempting to start a project, which could lead to cryptic errors. When supplying config value flags to `ddev config`, no validation was done before writing the config file. 

## How this PR Solves The Problem:
These changes do more to ensure that configuration values are valid before either 1) writing a config file, or 2) attempting to start a project. Additionally, many string values in the source code were constantized and a few helper functions were added as a requirement for valid value lookup.

Similarly, instead of inspecting error values, a small set of error types has been added allowing for error type comparisons, decoupling the error from its text value.

## Manual Testing Instructions:
Using the CLI's config flags, ensure that invalid choices lead to validation errors:

- `--projectname invalid!name`
- `--php-version 1.9`
- `--projecttype typo4`
- `--webserver-type nginx-cgi`
- `--additional-hostnames valid,in_valid`
- `--additional-fqdns valid.com,in_valid.com`

Provide combinations of these invalid values alongside valid values to ensure ddev rejects invalid values and, when appropriate, outputs the valid values for that flag.

Using config.yaml, ensure that invalid values lead to validation errors after saving the file and when attempting to `ddev start` the project:

- `name: "invalid name"`
- `type: drupal10`
- `php-version: 9.9`
- `webserver-type: nginx-cgi`
- `additional-hostnames: [valid, in_valid]`
- `additional-fqdns: [valid.com, in_valid.com]`
- `provider: unknown`

## Automated Testing Overview:
Constantization updates will largely rely on the build process completing successfully, ensuring all new 
symbols are resolved correctly.

Test cases have been added to ensure invalid config values generated appropriate errors.

## Related Issue Link(s):
#1095 

## Release/Deployment notes:
